File: scripts/algorithms/F/Find the Shortest Superstring/Find the Shortest Superstring.py
// Runtime: 551 ms (Top 76.67%) | Memory: 73.00 MB (Top 16.67%)

class Solution:
    def shortestSuperstring(self, words: List[str]) -> str:
        n = len(words)
        graph = [[0]*n for _ in range(n)] # graph as adjacency matrix 
        
        for i in range(n):
            for j in range(n): 
                if i != j: 
                    for k in range(len(words[j])): 
                        if words[i].endswith(words[j][:k]): 
                            graph[i][j] = len(words[j]) - k 
                            
        @cache
        def fn(prev, mask): 
            """Return length of shortest superstring & current choice of word."""
            if mask == 0: return 0, None
            vv, kk = inf, None
            for k in range(n): 
                if mask & 1<<k: 
                    v, _ = fn(k, mask ^ 1<<k)
                    offset = len(words[k]) if prev == -1 else graph[prev][k]
                    if v + offset < vv: vv, kk = v + offset, k
            return vv, kk
        
        ans = []
        prev = -1 
        mask = (1<<n) - 1
        while mask: 
            _, k = fn(prev, mask)
            if ans: ans.append(words[k][-graph[prev][k]:])
            else: ans.append(words[k])
            prev = k
            mask ^= 1<<k 
        return "".join(ans)

File: scripts/algorithms/F/Find the Shortest Superstring/Find the Shortest Superstring.rs
// Runtime: 20 ms (Top 100.0%) | Memory: 3.50 MB (Top 66.67%)

impl Solution {
    pub fn shortest_superstring(words: Vec<String>) -> String {
        let n = words.len();
        let mut graph = vec![vec![0; n]; n];
        for i in 0..n {
            for j in 0..n {
                if i != j {
                    let mut k = words[j].len();
                    while !words[i].ends_with(&words[j][..k]) {
                        k -= 1;
                    }
                    graph[i][j] = k;
                }
            }
        }
        let mut dp = vec![vec![0; n]; 1 << n];
        let mut parent = vec![vec![None; n]; 1 << n];
        for mask in 0..1 << n {
            for b in 0..n {
                if mask >> b & 1 == 0 {
                    continue;
                }
                let prev = mask ^ 1 << b;
                for (i, row) in graph.iter().enumerate() {
                    if prev >> i & 1 == 0 {
                        continue;
                    }
                    let val = dp[prev][i] + row[b];
                    if val > dp[mask][b] {
                        dp[mask][b] = val;
                        parent[mask][b] = Some(i);
                    }
                }
            }
        }
        let mut mask = (1 << n) - 1;
        let mut perm = Vec::with_capacity(n);
        let mut seen = vec![false; n];
        if let Some(mut p) = (0..n).max_by_key(|&i| dp[(1 << n) - 1][i]) {
            loop {
                perm.push(p);
                seen[p] = true;
                if let Some(t) = parent[mask][p] {
                    mask ^= 1 << p;
                    p = t;
                } else {
                    break;
                }
            }
        }
        perm.extend((0..n).filter(|&i| !seen[i]));
        (1..perm.len())
            .rev()
            .fold(words[perm[n - 1]].clone(), |acc, i| {
                let overlap = graph[perm[i]][perm[i - 1]];
                acc + &words[perm[i - 1]][overlap..]
            })
    }
}
